### PR TITLE
Add contributing document

### LINKED ISSUE #429: Consider adding CONTRIBUTING.md?
It would be nice to answer and explain the following in a public document like `CONTRIBUTING.md`:

1. Can I propose a new feature to this repo?
2. How do I submit a bug report?
3. What contributions are accepted?
4. Are there any special code organization rules I should know about?
5. What's the commit style (if any)?
6. Where can I suggest additions to `devcontainer-feature.json`?
7. What code formatter (if any) should I use?
8. How do I open a PR?
9. How do I get started testing these features?
10. How do I get started making changes?

These are all good questions that beginner contributors (like me) will inevitably ask. Having an encouraging document with a loose guide for some common things would be beneficial, I think.

I see there is a small "Contributing" section at the bottom of the readme, but I think more elaboration and detail would be good. And there's also the fact that some people won't scroll down to actually read it, but will see an all-caps "CONTRIBUTING.md" file and read that. 🤷‍♂️ It happens.

### PR DESCRIPTION
**Pull request by jcbhmr**
This is a first draft of a CONTRIBUTING.md. You can preview what it looks like here: https://github.com/jcbhmr/features/blob/add-contributing/CONTRIBUTING.md

It doesn't contain answers to every question in #429. Let me know if I should include more blurbs, sections, or other references to answer these questions. I think the current draft includes the most important information?

Additional questions:
1. Should any more details about the CLA process be added to the document?
2. What should go in the short security section? Is what I have there OK? I stole it right from https://github.com/devcontainers/spec/blob/main/SECURITY.md

Closes #429

### CODE REVIEW

**Comment by jcbhmr** [comment]
@microsoft-github-policy-service agree

**Comment by samruddhikhandale** [comment]
Beautiful, thank you so much for putting this up! ✨ 
I agree having a CONTRIBUTING.md would be helpful. 

@bamurtaugh / @Chuxel before I jump on to reviewing this PR, would appreciate your insights. Do we have a design on how we should structure this file? or any other guidelines would be helpful.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+- **❓ I have a question:** [Open an Issue]
+- **🐛 I found a bug:** [Open an Issue]
+- **💡 I have an idea:** [Open an Issue]
+- **💻 I want to code:** See below
+
+If you like the project, but just don't have time to contribute, that's OK too!
+You can also star the project ⭐, rave about it online 💬, or add a link to us
+🔗 in your project's readme.
+
+⚠️ You must never report security 🔒 related issues, vulnerabilities or bugs
```
Comment (bamurtaugh) [inline CONTRIBUTING.md:15]: I'm not sure if we need this. If it's a Microsoft-related security issue, this makes sense, but security issues may be beyond Microsoft as dev containers are more agnostic.

cc @chrmarti @alexdima for any input

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+including sensitive information to the issue tracker, or elsewhere in public.
+Instead, please [report them to the Microsoft Security Response Center]. You can
+read more about our security policy in [`SECURITY.md`].
+
+## Contributing code
+
+👨‍💼 This repository is a collection of devcontainer features managed by
```
Comment (bamurtaugh) [inline CONTRIBUTING.md:22]: We refer to dev containers as 2 words, so it'd be fantastic if you could update this PR to refer to them as "dev containers" 😄

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+devcontainer spec maintainers. Even though **we do not accept new features in
```
Comment (bamurtaugh) [inline CONTRIBUTING.md:23]: Also wording nit: We capitalize "Features" 😄

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+this repository**, we welcome contributions to [our existing features]!
+
+🚀 If you're looking to create a new feature, we encourage you to [create your
```
Comment (bamurtaugh) [inline CONTRIBUTING.md:26]: Would recommend linking to the feature-starter repo: https://github.com/devcontainers/feature-starter.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+1. 🔀 Fork the repo
+2. 💻 Open the repo in your editor
+3. 👨‍💻 Add your changes to your workspace
+4. ✨ Run the tests to make sure everything works
+5. 🔖 Commit & push your changes
+6. 🔁 Open a PR to get your changes merged
+7. 🚀 Profit!
+
+👩‍⚖️ When contributing code to this project, you may be asked to agree to our
+[Contributor License Agreement].
```
Comment (bamurtaugh) [inline CONTRIBUTING.md:41]: Also not sure about this for potentially more agnostic contributions, cc @chrmarti @alexdima

**Review by bamurtaugh — COMMENTED** [review]
Thanks so much @jcbhmr for taking the time to put together a fantastic doc! Left a few comments and suggestions.

@samruddhikhandale I think this looks great for a basis we can build up and iterate on over time, and definitely want to hear your feedback if you have other thoughts.

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,52 @@
+**Thanks for taking the time to contribute! ❤️**
+
+All types of contributions are encouraged and valued, no matter if it's a bug
+report 🐛, a feature request 💡, or a Pull Request 🚀.
+
+- **❓ I have a question:** [Open an Issue]
```
[the remark this answers is not in the document]
Comment (jcbhmr) [reply CONTRIBUTING.md]: Done in https://github.com/jcbhmr/features/commit/8b585825b220178c9f9a1ca2fa815f1551d011b9

File: CONTRIBUTING.md (hunk 4 above)
Comment (jcbhmr) [reply CONTRIBUTING.md:26]: Is it sufficient to link here?

image

File: CONTRIBUTING.md (hunk 2 above)
Comment (jcbhmr) [reply CONTRIBUTING.md]: Fixed in https://github.com/jcbhmr/features/commit/d7861eee339f77ce5b5762a3fe84baed642ecaca

File: CONTRIBUTING.md (hunk 3 above)
Comment (jcbhmr) [reply CONTRIBUTING.md]: Fixed in https://github.com/jcbhmr/features/commit/d7861eee339f77ce5b5762a3fe84baed642ecaca

File: CONTRIBUTING.md (hunk 5 above)
Comment (chrmarti) [reply CONTRIBUTING.md:41]: Replied more generally above.

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+⚠️ You must never report security 🔒 related issues, vulnerabilities or bugs
+including sensitive information to the issue tracker, or elsewhere in public.
+Instead, please [report them to the Microsoft Security Response Center]. You can
+read more about our security policy in [`SECURITY.md`].
+
+## Contributing code
+
+👨‍💼 This repository is a collection of dev container Features managed by dev
+container spec maintainers. Even though **we do not accept new Features in this
+repository**, we welcome contributions to [our existing Features]!
```
Comment (samruddhikhandale) [inline CONTRIBUTING.md:24]:
```suggestion
👨‍💼 This repository is a collection of dev container Features managed by dev
container spec maintainers. This repository will **only accept improvements and bug fixes** for the [current set of maintained Features](https://github.com/devcontainers/features/blob/main/src)
```

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+🚀 If you're looking to create a new Feature, we encourage you to [create your
+own Feature repository] and add it to the index in [`collection-index.yml`].
```
Comment (samruddhikhandale) [inline CONTRIBUTING.md:27]:
```suggestion
🚀 If you're looking to create a new Feature, then we encourage adding it to a separate repository of your control. The [devcontainers/feature-starter](https://github.com/devcontainers/feature-starter) repository has more guidance on self authoring Features.

💡 Once you've self authored the Features and if you'd like to share them with the community, then we'd recommend [adding it to the index](https://github.com/devcontainers/feature-starter#adding-features-to-the-index).
```

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
+👷‍ If you've identified an issue and you want to fix it, here's how you can get
+started:
+
+1. 🔀 Fork the repo
+2. 💻 Open the repo in your editor
+3. 👨‍💻 Add your changes to your workspace
+4. ✨ Run the tests to make sure everything works
```
Comment (samruddhikhandale) [inline CONTRIBUTING.md:35]: Can we add a reference link to https://github.com/devcontainers/cli/blob/main/docs/features/test.md? thanks!

File: CONTRIBUTING.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (jcbhmr) [reply CONTRIBUTING.md]: Which bullets do you want mapped to which links? Should _all_ be mapped to "ask in our slack channel?" 🤔 Right now it's

- question = slack
- bug = issue
- idea = issue
- code = pr

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+5. 🔖 Commit & push your changes
+6. 🔁 Open a PR to get your changes merged
```
Comment (samruddhikhandale) [inline CONTRIBUTING.md:37]: Can we add a note that every PR should include a change to the `version` field of the updated Feature? It should be bumped by following semantic versioning.

Sample - x.y.x
x: major version - Breaking change/ Backwards compat
y: minor version - Adding a new Feature option/capability
z: patch version - Bug fixes

File: CONTRIBUTING.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (samruddhikhandale) [reply CONTRIBUTING.md]: Oops! Pasted an incorrect screenshot.
Referring to `question = slack`

File: CONTRIBUTING.md (hunk 8 above)
Comment (jcbhmr) [reply CONTRIBUTING.md:27]: Fixed in https://github.com/jcbhmr/features/commit/1a50ace247e56b9186609e9d76bef24666295bac

File: CONTRIBUTING.md (hunk 7 above)
Comment (jcbhmr) [reply CONTRIBUTING.md]: Fixed in https://github.com/jcbhmr/features/commit/1a50ace247e56b9186609e9d76bef24666295bac

File: CONTRIBUTING.md (hunk 8 above)
Comment (jcbhmr) [reply CONTRIBUTING.md]: Fixed in https://github.com/jcbhmr/features/commit/b3f4b8b975087af2029bea3581dcb15747bbbb74

File: CONTRIBUTING.md (hunk 9 above)
Comment (jcbhmr) [reply CONTRIBUTING.md]: Fixed in https://github.com/jcbhmr/features/commit/3aa8264f0cabb849ba3487028a27803a67cac91d

File: CONTRIBUTING.md (hunk 4 above)
Comment (jcbhmr) [reply CONTRIBUTING.md]: Fixed in https://github.com/jcbhmr/features/commit/b3f4b8b975087af2029bea3581dcb15747bbbb74

File: CONTRIBUTING.md (hunk 10 above)
Comment (jcbhmr) [reply CONTRIBUTING.md]: Like this? https://github.com/jcbhmr/features/commit/b16118266583ed22626e8fe7ef68be091c491a4d

File: CONTRIBUTING.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (jcbhmr) [reply CONTRIBUTING.md]: Fixed in https://github.com/jcbhmr/features/commit/b1f61abb2f69d9f14ee4d511d046c08c8b064572

**Review by samruddhikhandale — DISMISSED** [review]
This is awesome, thank you! 🚀

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+1. 🔀 Fork the repo
+2. 💻 Open the repo in your editor
+3. 👨‍💻 Add your changes to your workspace
+4. ✨ [Test your changes using `devcontainer features test`] to make sure
+   everything still works
+5. 🆚 Bump the version of the feature you changed according to [semver]
+6. 🔖 Commit & push your changes
+7. 🔁 Open a PR to get your changes merged
+8. 🚀 Profit!
+
```
[the remark this answers is not in the document]
Comment (jcbhmr) [reply CONTRIBUTING.md:45]: This OK? Or did you want to remove/reformat this part more? https://github.com/jcbhmr/features/commit/27baabc871e8d7620434df6cfa58efcb50af5b1f

image

[the remark this answers is not in the document]
Comment (samruddhikhandale) [reply CONTRIBUTING.md:45]: Looks good!

**Review by samruddhikhandale — APPROVED** [review]

**Review by Genio40921 — COMMENTED** [review]
Fin

**Comment by mtrackeros** [comment]
> This is a first draft of a CONTRIBUTING.md. You can preview what it looks like here: https://github.com/jcbhmr/features/blob/add-contributing/CONTRIBUTING.md
> 
> It doesn't contain answers to every question in #429. Let me know if I should include more blurbs, sections, or other references to answer these questions. I think the current draft includes the most important information?
> 
> Additional questions:
> 
> 1. Should any more details about the CLA process be added to the document?
> 2. What should go in the short security section? Is what I have there OK? I stole it right from https://github.com/devcontainers/spec/blob/main/SECURITY.md
> 
> Closes #429

**Comment by mtrackeros** [comment]
> This is a first draft of a CONTRIBUTING.md. You can preview what it looks like here: https://github.com/jcbhmr/features/blob/add-contributing/CONTRIBUTING.md
> 
> It doesn't contain answers to every question in #429. Let me know if I should include more blurbs, sections, or other references to answer these questions. I think the current draft includes the most important information?
> 
> Additional questions:
> 
> 1. Should any more details about the CLA process be added to the document?
> 2. What should go in the short security section? Is what I have there OK? I stole it right from https://github.com/devcontainers/spec/blob/main/SECURITY.md
> 
> Closes #429

**Pull request merged**